frontend/src/components/UserGroupCard/index.tsx

```diff
@@ -1,53 +1,88 @@
 import React, { useEffect, useState } from "react";
 import "./styles.css";
-import { fetchUsersWithRole } from "@/api/users";
+import { fetchUser, fetchUsersWithRole } from "@/api/users";
 
 interface IUserGroupCardProps {
   label: string;
   role_type: string;
   classroom: IClassroom;
+  givenUsersList?: IClassroomUser[];
   onClick?: () => void;
 }
 
 const UserGroupCard: React.FC<IUserGroupCardProps> = ({
   label,
   role_type,
   classroom,
+  givenUsersList,
   onClick,
 }) => {
-  const [numUsers, setNumUsers] = useState<number>(0);
-
+  const [numUsers, setNumUsers] = useState<number>(
+    givenUsersList ? givenUsersList.length : 0
+  );
+  const [userMap, setUserMap] = useState<Map<IClassroomUser, IGitHubUser>>(
+    new Map()
+  );
   useEffect(() => {
-    const getUsers = async () => {
-      try {
-        const users = await fetchUsersWithRole(role_type, classroom);
-        setNumUsers(users.length);
-      } catch (error) {
-        console.error("Error fetching users:", error);
+    const loadGitHubUsers = async () => {
+      if (givenUsersList) {
+        const newMap = new Map();
+        await Promise.all(
+          givenUsersList.map(async (classroomUser) => {
+            await fetchUser(classroomUser.github_username)
+              .then((userResponse: IGitHubUserResponse) => {
+                newMap.set(classroomUser, userResponse.user);
+              })
+              .catch((error) => {
+                console.error(`Error fetching GitHub user:`, error);
+              });
+          })
+        );
+        setUserMap(newMap);
       }
     };
 
-    void getUsers();
-  }, [role_type, classroom]);
+    void loadGitHubUsers();
+  }, [givenUsersList]);
 
-  let userIcons = [];
+  useEffect(() => {
+    if (givenUsersList) {
+      setNumUsers(givenUsersList.length);
+    } else {
+      const getUsers = async () => {
+        await fetchUsersWithRole(role_type, classroom)
+          .then((users) => {
+            setNumUsers(users.length);
+          })
+          .catch((error) => {
+            console.error("Error fetching users:", error);
+          });
+      };
```

Comment:
Just jumped ahead to the dashboard page that uses this component. Is it bad practice to be fetching users here while already having fetched it there? Could this lead to inconsistent states between the components?

Reply:
these are actually two separate operations. One gets our database's representation of the class roster. Then here, we are querying for Github's representation of each student, that way we can display their user icon.